server.go

```diff
@@ -998,16 +1009,20 @@ func (s *server) peerTerminationWatcher(p *peer) {
 		}
 	}
 
-	// Send the peer to be garbage collected by the server.
-	s.removePeer(p)
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
-	// If this peer had an active persistent connection request, then we
-	// can remove this as we manually decide below if we should attempt to
-	// re-connect.
-	if p.connReq != nil {
-		s.connMgr.Remove(p.connReq.ID())
+	// If the server has already removed this peer, we can short circuit the
+	// peer termination watcher and skip cleanup.
+	if _, ok := s.ignorePeerTermination[p]; ok {
+		delete(s.ignorePeerTermination, p)
+		return
 	}
 
+	// First, cleanup any remaining state the server has regarding the peer
+	// in question.
+	s.removePeer(p)
+
 	// Next, check to see if this is a persistent peer or not.
```

Comment:
It looks like this reconnect logic in gets skipped if disconnected by us. Is that always OK?

Reply:
There are currently three places where we can intentionally disconnect a peer. All persistent conn reqs are cleaned up during all three cases, so no reconnection should be attempted for any of them.

The first is through the server's `DisconnectPeer` method, in which the user (or autopilot i suppose) has instructed us that we don't need to communicate anymore.

The second and third happen in InboundPeerConnected and OutboundPeerConnected. In either case, we may need to disconnect an existing peer because we will be atomically swapping in a new valid peer for the same counterparty/pubkey. We do not need to attempt any reconnects for the old peer, since we are keeping a new valid peer in its place.

Any other disconnects will be treated as unexpected by server, and the peerTerminationWatcher will be activated and attempt reconnection if it is required.

I think that covers how the system is intended to work, let me know if I can provide more clarification :)

Reply:
One clarification: the `persistentPeers` map is unaltered when hot swapping in peers coming from inbound or outbound connections. Thus any entries in the `persistentPeers` map will propagate over to the new peer, and an reconnection will be attempted if the new peer unexpectedly disconnects.